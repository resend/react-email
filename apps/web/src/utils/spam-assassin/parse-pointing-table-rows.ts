export const parsePointingTableRows = (response: string) => {
  const tableHeader =
    /pts\s+rule\s+name\s+description\s+(?<ptsWidth>-+) (?<ruleNameWidth>-+) (?<descriptionWidth>-+) *(?:\r\n|\n|\r)*/;
  const tableStartMatch = response.match(tableHeader);

  if (
    tableStartMatch === null ||
    tableStartMatch.index === undefined ||
    tableStartMatch.groups === undefined
  ) {
    throw new Error('Could not find spam checking points table');
  }

  const ptsWidth = tableStartMatch.groups.ptsWidth!.length;
  const ruleNameWidth = tableStartMatch.groups.ruleNameWidth!.length;
  const descriptionWidth = tableStartMatch.groups.descriptionWidth!.length;
  const columnsRegex = new RegExp(
    `^(?<pts>.{${ptsWidth}}) (?<ruleName>.{${ruleNameWidth}}) (?<description>.{${descriptionWidth}}|.+$)`,
  );

  const rows: {
    pts: number;
    ruleName: string;
    description: string;
  }[] = [];

  const responseFromTableStart = response.slice(
    tableStartMatch.index + tableStartMatch[0].length,
  );
  for (const line of responseFromTableStart.split(/\r\n|\n|\r/)) {
    if (line.trim().length === 0) break;

    const match = line.match(columnsRegex);
    if (match?.groups === undefined) {
      throw new Error('Could not match the columns in the row', {
        cause: {
          line,
          match,
        },
      });
    }
    const pts = match.groups.pts!;
    const ruleName = match.groups.ruleName!;
    const description = match.groups.description!;

    try {
      rows.push({
        pts: Number.parseFloat(pts),
        ruleName: ruleName.trim(),
        description: description.trim(),
      });
    } catch (exception) {
      throw new Error('could not parse points to insert into rows array', {
        cause: {
          exception,
          line,
          match,
        },
      });
    }
  }

  return rows;
};
