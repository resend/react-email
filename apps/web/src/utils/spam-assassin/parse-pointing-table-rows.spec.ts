import { parsePointingTableRows } from './parse-pointing-table-rows';

describe('parsePointingTableRows()', () => {
  test('works with spammy emails', () => {
    const spamdResponse = `Received: from localhost by gabriels-computer
        with SpamAssassin (version 4.0.1);
        Mon, 10 Feb 2025 09:21:23 -0300
X-Spam-Checker-Version: SpamAssassin 4.0.1 (2024-03-26) on gabriels-computer
X-Spam-Flag: YES
X-Spam-Level: **********
X-Spam-Status: Yes, score=10.2 required=5.0 tests=DRUGS_ERECTILE,HTML_MESSAGE,
        MISSING_DATE,MISSING_FROM,MISSING_HEADERS,MISSING_MID,MISSING_SUBJECT,
        MONEY_BACK,NO_HEADERS_MESSAGE,NO_RECEIVED,NO_RELAYS autolearn=no
        autolearn_force=no version=4.0.1
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="----------=_67A9EF43.EC247F5D"

This is a multi-part message in MIME format.

------------=_67A9EF43.EC247F5D
Content-Type: text/plain; charset=UTF-8
Content-Disposition: inline
Content-Transfer-Encoding: 8bit

Spam detection software, running on the system "gabriels-computer",
has identified this incoming email as possible spam.  The original
message has been attached to this so you can view it or label
similar future email.  If you have any questions, see
root@localhost for details.

Content preview:  This email is spam. We sell rolexss for cheap. Get your viagra.
   We also sell weight loss pills today. Money back guaranteed. sEnd me $500
   and I'll send you back $700. don't tell anyone. Send this emai [...] 

Content analysis details:   (10.2 points, 5.0 required)

 pts rule name              description
---- ---------------------- --------------------------------------------------
 1.8 MISSING_SUBJECT        Missing Subject: header                           
 1.4 MISSING_DATE           Missing Date: header                              
 0.1 MISSING_MID            Missing Message-Id: header                        
 1.0 MISSING_FROM           Missing From: header                              
-0.0 NO_RECEIVED            Informational: message has no Received headers    
 1.2 MISSING_HEADERS        Missing To: header                                
-0.0 NO_RELAYS              Informational: message was not relayed via SMTP   
 2.5 MONEY_BACK             BODY: Money back guarantee                        
 0.0 HTML_MESSAGE           BODY: HTML included in message
 0.0 NO_HEADERS_MESSAGE     Message appears to be missing most RFC-822 headers
 2.2 DRUGS_ERECTILE         Refers to an erectile drug

The original message was not completely plain text, and may be unsafe to
open with some email clients; in particular, it may contain a virus,
or confirm that your address can receive spam.  If you wish to view
it, it may be safer to save it to a file and open it with an editor.


------------=_67A9EF43.EC247F5D
Content-Type: message/rfc822; x-spam-type=original
Content-Description: original message before SpamAssassin
Content-Disposition: attachment
Content-Transfer-Encoding: 8bit

MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="Part_af7eace217e10ccb689dd03f0d264877"

--Part_af7eace217e10ccb689dd03f0d264877
Content-Type: text/plain; charset="UTF-8"

This email is spam. We sell rolexss for cheap. Get your viagra. We also
sell weight loss pills today. Money back guaranteed. sEnd me $500 and
I'll send you back $700. don't tell anyone. Send this email to ten
friends

--Part_af7eace217e10ccb689dd03f0d264877
Content-Type: text/html; charset="UTF-8"

<html>
  <body>
    What
  </body>
</html>

--Part_af7eace217e10ccb689dd03f0d264877--


------------=_67A9EF43.EC247F5D--
`;
    expect(parsePointingTableRows(spamdResponse)).toMatchSnapshot();
  });

  test('works with a multiline description', () => {
    const partialSpamResponse = `â€Œâ [...]

Content analysis details:   (9.4 points, 5.0 required)

 pts rule name              description
---- ---------------------- --------------------------------------------------
-0.0 NO_RECEIVED            Informational: message has no Received headers
 0.1 MISSING_MID            Missing Message-Id: header
 1.4 MISSING_DATE           Missing Date: header
 1.0 MISSING_FROM           Missing From: header
 1.8 MISSING_SUBJECT        Missing Subject: header
 1.2 MISSING_HEADERS        Missing To: header
-0.0 NO_RELAYS              Informational: message was not relayed via SMTP
 0.0 URIBL_BLOCKED          ADMINISTRATOR NOTICE: The query to URIBL was blocked.
                            See
                            http://wiki.apache.org/spamassassin/DnsBlocklists#dnsbl-block
                             for more information.
                            [URI: stripe.com]
 0.0 HTML_MESSAGE           BODY: HTML included in message
 0.0 NO_HEADERS_MESSAGE     Message appears to be missing most RFC-822 headers
 3.9 DOS_BODY_HIGH_NO_MID   High bit body and no message ID header

The original message was not completely plain text, and may be unsafe to
open with some email clients; in particular, it may contain a virus,
or confirm that your address can receive spam.  If you wish to view
it, it may be safer to save it to a file and open it with an editor.


------------=_67BF2F0C.CEF6CDE8
Content-Type: message/rfc822; x-spam-type=original`;

    expect(parsePointingTableRows(partialSpamResponse)).toMatchSnapshot();
  });
});
