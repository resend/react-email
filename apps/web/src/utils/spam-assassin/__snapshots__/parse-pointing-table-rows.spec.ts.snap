// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parsePointingTableRows() > works with a multiline description 1`] = `
[
  {
    "description": "Informational: message has no Received headers",
    "pts": -0,
    "ruleName": "NO_RECEIVED",
  },
  {
    "description": "Missing Message-Id: header",
    "pts": 0.1,
    "ruleName": "MISSING_MID",
  },
  {
    "description": "Missing Date: header",
    "pts": 1.4,
    "ruleName": "MISSING_DATE",
  },
  {
    "description": "Missing From: header",
    "pts": 1,
    "ruleName": "MISSING_FROM",
  },
  {
    "description": "Missing Subject: header",
    "pts": 1.8,
    "ruleName": "MISSING_SUBJECT",
  },
  {
    "description": "Missing To: header",
    "pts": 1.2,
    "ruleName": "MISSING_HEADERS",
  },
  {
    "description": "Informational: message was not relayed via SMTP",
    "pts": -0,
    "ruleName": "NO_RELAYS",
  },
  {
    "description": "ADMINISTRATOR NOTICE: The query to URIBL was blocked. See http://wiki.apache.org/spamassassin/DnsBlocklists#dnsbl-block for more information. [URI: stripe.com]",
    "pts": 0,
    "ruleName": "URIBL_BLOCKED",
  },
  {
    "description": "BODY: HTML included in message",
    "pts": 0,
    "ruleName": "HTML_MESSAGE",
  },
  {
    "description": "Message appears to be missing most RFC-822 headers",
    "pts": 0,
    "ruleName": "NO_HEADERS_MESSAGE",
  },
  {
    "description": "High bit body and no message ID header",
    "pts": 3.9,
    "ruleName": "DOS_BODY_HIGH_NO_MID",
  },
]
`;

exports[`parsePointingTableRows() > works with spammy emails 1`] = `
[
  {
    "description": "Missing Subject: header",
    "pts": 1.8,
    "ruleName": "MISSING_SUBJECT",
  },
  {
    "description": "Missing Date: header",
    "pts": 1.4,
    "ruleName": "MISSING_DATE",
  },
  {
    "description": "Missing Message-Id: header",
    "pts": 0.1,
    "ruleName": "MISSING_MID",
  },
  {
    "description": "Missing From: header",
    "pts": 1,
    "ruleName": "MISSING_FROM",
  },
  {
    "description": "Informational: message has no Received headers",
    "pts": -0,
    "ruleName": "NO_RECEIVED",
  },
  {
    "description": "Missing To: header",
    "pts": 1.2,
    "ruleName": "MISSING_HEADERS",
  },
  {
    "description": "Informational: message was not relayed via SMTP",
    "pts": -0,
    "ruleName": "NO_RELAYS",
  },
  {
    "description": "BODY: Money back guarantee",
    "pts": 2.5,
    "ruleName": "MONEY_BACK",
  },
  {
    "description": "BODY: HTML included in message",
    "pts": 0,
    "ruleName": "HTML_MESSAGE",
  },
  {
    "description": "Message appears to be missing most RFC-822 headers",
    "pts": 0,
    "ruleName": "NO_HEADERS_MESSAGE",
  },
  {
    "description": "Refers to an erectile drug",
    "pts": 2.2,
    "ruleName": "DRUGS_ERECTILE",
  },
]
`;
